### PR TITLE
Ghost teleport support/ Medevac timer

### PR DESCRIPTION
**Pull request by Redory**
## About The Pull Request

Added ghost teleport support for:
- Medevac
- Xeno tunnels
- ASRS teleporter
- Wraith portal
- Shuttle computer

While redoing the medevac linking system i implemented a simple timer onto the medevac roller so you can see the cooldown in realtime in your inventory:
<h5>(No, that is not a stack of medevacs in your hand, i swear!)</h5>

image

Inspecting a medevac beacon also shows the status of the medevac stretchers and who is currently carrying them:
image

Most of the code is copied from a PR i made 2 years ago: #10622 

## Why It's Good For The Game

Ghost man teleport gud.
WASD around and orbiting random people to get to ship/ teleporter bad.
I AM A GHOOOOST AND WISH TO WHOOOSH AROUND FASTER
## Changelog
:cl:
qol: Ghosts can teleport using Medevacs, Xeno tunnels, ASRS teleporters, Wraith portals and Shuttle computers.
qol: Medevac stretchers now have a realtime cooldown on their sprite when in inventory.
qol: Medevac beacons now list all linked stretchers and show their cooldowns.
/:cl:

### CODE REVIEW

**Comment by lbnesquik** [comment]
Does that mean ghosts can click to see where the medivac bed goes?

**Comment by Redory** [comment]
> Does that mean ghosts can click to see where the medivac bed goes?

yes

File: code/modules/shuttle/marine_dropship.dm [hunk 1]
```diff
@@ -1270,15 +1270,16 @@
...
 		RelinkShuttleId()
 	ui = SStgui.try_update_ui(user, src, ui)
 	if(!ui)
 		ui = new(user, src, "ShuttleControl")
 		ui.open()
 
 /obj/machinery/computer/shuttle/shuttle_control/ui_state(mob/user)
-	return GLOB.access_state
+	var/datum/ui_state/include_observer_state/observer_state = GLOB.include_observer_state
+	return new observer_state(GLOB.access_state)
```
Comment (ivanmixo) [inline code/modules/shuttle/marine_dropship.dm:1282]: Uh.. no new. Look at how the other states are done.

File: code/game/objects/structures/teleporter.dm [hunk 2]
```diff
@@ -93,6 +93,12 @@
 	for(var/atom/movable/thing_to_teleport AS in teleporting)
 		thing_to_teleport.forceMove(get_turf(deployed_linked_teleporter))
 
+/obj/machinery/deployable/teleporter/attack_ghost(mob/dead/observer/user)
+	var/obj/item/teleporter_kit/kit = internal_item
+	if(!kit || !kit.linked_teleporter)
```
Comment (ivanmixo) [inline code/game/objects/structures/teleporter.dm:98]: Replace !kit with !istype(kit) so you get a null check and type check in one.

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 3]
```diff
@@ -626,6 +662,32 @@ GLOBAL_LIST_EMPTY(activated_medevac_stretchers)
 		icon_state = "med_beacon0"
 		playsound(loc,'sound/machines/click.ogg', 25, FALSE)
 
+/obj/item/medevac_beacon/attack_ghost(mob/dead/observer/user)
+	. = ..()
+	if(!linked_beds && !linked_beds_deployed)
+		return
+	var/list/obj/destinations = SANITIZE_LIST(linked_beds) + SANITIZE_LIST(linked_beds_deployed)
+	var/obj/O
+	if(LAZYLEN(linked_beds + linked_beds_deployed) > 1)
```
Comment (ivanmixo) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:671]:
```suggestion
	if(length(linked_beds + linked_beds_deployed) > 1)
```

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 4]
```diff
@@ -643,15 +705,65 @@ GLOBAL_LIST_EMPTY(activated_medevac_stretchers)
...
 /obj/item/medevac_beacon/proc/check_power()
 	var/area/A = loc?.loc
 	if(!A || !isarea(A))
 		return FALSE
 	return(A.powered(1))
+
+/// Adds a medevac roller or stretcher to the medevac beacon. Returns TRUE if the beacon is in the linked_beds* list and false if it is not in there.
+/obj/item/medevac_beacon/proc/add_stretcher(obj/O, mob/user, silent = FALSE)
+	var/obj/item/roller/medevac/R = O
+	if(R && istype(R, /obj/item/roller/medevac))
```
Comment (ivanmixo) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:720]: istype should account for there not being an `R`. Also chill with the single letter var names, make them descriptive.

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 5]
```diff
@@ -530,29 +558,36 @@ GLOBAL_LIST_EMPTY(activated_medevac_stretchers)
...
 
 	. += span_notice("[details.Join(" ")]")
 
+/obj/item/roller/medevac/update_overlays()
+	. = ..()
+	var/display_timer_cooldown = CEILING((timer_cooldown) * 0.1, 1)
+	if(isturf(loc) || !display_timer_cooldown)
+		return
+	var/mutable_appearance/desc = mutable_appearance('icons/misc/12x12.dmi')
+	desc.maptext = MAPTEXT(display_timer_cooldown)
```
Comment (ivanmixo) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:571]: Add an 's' to it so it's indicative of it being seconds and not it looking like a stack.

**Review by ivanmixo — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 6]
```diff
@@ -643,15 +704,65 @@ GLOBAL_LIST_EMPTY(activated_medevac_stretchers)
...
 /obj/item/medevac_beacon/proc/check_power()
 	var/area/A = loc?.loc
 	if(!A || !isarea(A))
 		return FALSE
 	return(A.powered(1))
+
+/// Adds a medevac roller or stretcher to the medevac beacon. Returns TRUE if the beacon is in the linked_beds* list and false if it is not in there.
+/obj/item/medevac_beacon/proc/add_stretcher(obj/target_bed, mob/user, silent = FALSE)
+	var/obj/item/roller/medevac/rollerbed = target_bed
+	if(rollerbed && istype(rollerbed, /obj/item/roller/medevac))
```
Comment (ivanmixo) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:719]: Checking for rollerbed is not necesarry here, just keep the istype check. Same for line 737

File: code/modules/shuttle/marine_dropship.dm [hunk 7]
```diff
@@ -1346,6 +1346,32 @@
...
+	for(var/destination in all_destinations)
+		for(var/obj/docking_port/port AS in SSshuttle.stationary)
+			if(destination != port.id)
+				continue
+			port_assoc["[port.name]"] = destination
+
+	var/list/destinations = list()
+	for(var/destination in port_assoc)
+		destinations += destination
+	var/input = tgui_input_list(user ,"Choose a port to teleport to:" ,"Ghost Shuttle teleport" ,destinations ,null, 0)
```
Comment (ivanmixo) [inline code/modules/shuttle/marine_dropship.dm:1366]:
```suggestion
	var/input = tgui_input_list(user ,"Choose a port to teleport to:", "Ghost Shuttle teleport", destinations, null, 0)
```

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 8]
```diff
@@ -626,6 +661,32 @@ GLOBAL_LIST_EMPTY(activated_medevac_stretchers)
...
+	var/obj/target
+	if(length(linked_beds + linked_beds_deployed) > 1)
+		var/list/medevac_assoc = list()
+		for(var/obj/destination in destinations)
+			var/turf/T = get_turf(destination)
+			medevac_assoc["X:[T.x], Y:[T.y] - \[[get_area(destination)]\]"] = destination
+		destinations = list()
+		for(var/destination in medevac_assoc)
+			destinations += destination
+		var/input = tgui_input_list(user ,"Choose a medevac to teleport to:" ,"Ghost Medevac teleport" ,destinations ,null, 0)
```
Comment (ivanmixo) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:678]:
```suggestion
		var/input = tgui_input_list(user ,"Choose a medevac to teleport to:", "Ghost Medevac teleport", destinations, null, 0)
```

File: code/modules/xenomorph/xeno_structures.dm [hunk 9]
```diff
@@ -372,6 +372,33 @@ TUNNEL
...
+	var/list/obj/destinations = GLOB.xeno_tunnels_by_hive[hivenumber]
+	var/obj/structure/xeno/tunnel/targettunnel
+	if(LAZYLEN(destinations) > 2)
+		var/list/tunnel_assoc = list()
+		for(var/obj/D in destinations)
+			tunnel_assoc["X:[D.x], Y:[D.y] - \[[get_area(D)]\]"] = D
+		destinations = list()
+		for(var/d in tunnel_assoc)
+			destinations += d
+		var/input = tgui_input_list(user ,"Choose a tunnel to teleport to:" ,"Ghost Tunnel teleport" ,destinations ,null, 0)
```
Comment (ivanmixo) [inline code/modules/xenomorph/xeno_structures.dm:387]:
```suggestion
		var/input = tgui_input_list(user ,"Choose a tunnel to teleport to:", "Ghost Tunnel teleport", destinations, null, 0)
```

**Review by ivanmixo — CHANGES_REQUESTED** [review]
Few minor things left

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 10]
```diff
@@ -512,16 +513,43 @@ GLOBAL_LIST_EMPTY(activated_medevac_stretchers)
 	var/last_teleport = null
 	var/obj/item/medevac_beacon/linked_beacon = null
 	rollertype = /obj/structure/bed/medevac_stretcher
+	///Visual timer for the medevac cooldown
+	var/timer_cooldown
+	///Who is currently holding onto the medevac roller?
+	var/mob/holder
 
 /obj/item/roller/medevac/Destroy()
```
Comment (ivanmixo) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:521]: `holder` still needs to be nulled

**Review by ivanmixo — CHANGES_REQUESTED** [review]

**Review by ivanmixo — APPROVED** [review]

**Pull request merged**